Step3: 자동차 경주

wrallee opened the pull request:
안녕하세요.
3번째 미션 제출합니다.

지난번에 순수함수 이야기를 해주셔서 순수함수로 바꿔보기도 하고 추가적으로 Position, Car를 불변으로도 만들고 싶었는데요.
뭔가 파라미터가 많아지고 장황해지는것같기도 하고 상위레벨인 CarGame은 불변으로 하기 애매하다고 느껴졌습니다.

객체지향과 함수형 사이에 트레이드오프를 해야하는것인지..
아니면 제가 아직 함수형을 잘 이해를 못해서 구현이 어려운것인지 이번 미션을 진행하면서 궁금했습니다.

리뷰 부탁드리겠습니다.

wrallee committed:
test(RacingCar): 차 객체 테스트 작성

wrallee committed:
feat(RacingCar): 자동차 이동 기능 구현

wrallee committed:
feat(RacingCar): 대시보드 및 출력기능 구현

wrallee committed:
refactor(RacingCar): Position 객체 포장

wrallee committed:
refactor(RacingCar): Score 객체 포장

wrallee committed:
feat(RacingCar): 점수 랜덤함수 적용

wrallee committed:
refactor(RacingCar): ResultView로 출력로직 이동

kyucumber commented on src/main/kotlin/racingcar/model/Score.kt:
```diff
@@ -0,0 +1,19 @@
+private const val MINIMUM_SCORE = 0
+private const val MAXIMUM_SCORE = 9
+
+class Score(val value: Int) {
```
지금 이 클래스는 이동 여부를 결정하는 역할을 하고 있습니다. Score가 적절한 이름이 맞는지 조금 고민되네요.

wrallee replied:
MoveFactor로 변경했습니다.
이동하기 위해 랜덤 점수를 얻고, 이동여부는 각 Car의 Threshold값에 의해 결정된다고 생각했었는데 요구사항엔 없는 내용이라 적절하지 않았던걸까요?😶

kyucumber replied:
> 요구사항엔 없는 내용이라 적절하지 않았던걸까요?

그런 의미는 아니었어요. Score라는 이름 그대로 보면 점수를 저장하는 용도일 것 같은데 이동 여부를 결정하는 책임이 같이 포함되어 있어서 적절한 이름이 맞는지 고민된다는 내용의 코멘트였어요.

wrallee replied:
"점수"가 이동을 결정하는데 사용되는(책임이 있는)가를 고민해봤어야됐군요. 게임 애플리케이션이라 점수라는 도메인이 새로 생기는 케이스가 있을것같다는 생각도 들고..😅 이동 여부에 사용된다는 의미가 더 명확해지도록 네이밍하는게 좋은것 같습니다. 답변 감사합니다!

kyucumber commented on src/main/kotlin/racingcar/service/CarGame.kt:
```diff
@@ -0,0 +1,12 @@
+
+    fun play() {
+        cars.forEach { it.move(Score.random()) }
+    }
```
> 뭔가 파라미터가 많아지고 장황해지는것같기도 하고 상위레벨인 CarGame은 불변으로 하기 애매하다고 느껴졌습니다.

CarGame만 보았을 때 불변을 잘 유지하고 있는 것 같은데 어떤 부분에서 애매하다고 느끼셨는지 자세히 알려주실 수 있을까요?

wrallee replied:
CarGame은 불변으로 구현했더라도 내부의 Car.position이 가변이라 결국 CarGame도 가변이라 생각했었습니다.
그래서 Position의 increase 함수를 아래처럼 구현한다거나
```java
fun increase(speed: Int): Position {
    return Position(value + speed)
}
```

Car의 move 함수를 아래처럼 구현했었다가 롤백했었는데요.
```java
fun move(factor: MoveFactor): Car {
    if (factor.isBiggerThanOrEquals(MOVE_THRESHOLD)) {
        return Car(position.increase(DEFAULT_SPEED))
    }
    return Car(position)
}
```

CarGame도 play()를 호출할 때마다 새로운 CarGame을 리턴하도록 구현하려다보니 뭔가 너무 과해지는것같다는 느낌이 들어서 일단 롤백하고 Position만 가변으로 만든 뒤 이 내용은 따로 질문드린거였습니다. 롤백했던 히스토리는 쏙 빼고 질문드렸었네요ㅜ

그래도 아래에 답변을 해주셔서 잘 이해 된 것 같습니다!

kyucumber commented on src/main/kotlin/racingcar/model/Position.kt:
```diff
@@ -0,0 +1,14 @@
+
+private const val INITIAL_POSITION = 1
+
+class Position(var value: Int = INITIAL_POSITION) {
```
> 지난번에 순수함수 이야기를 해주셔서 순수함수로 바꿔보기도 하고 추가적으로 Position, Car를 불변으로도 만들고 싶었는데요.

순수함수나 불변을 사용했을 때 사이드이펙트 등을 줄이고 이해하기 쉬운 코드를 만들 수 있어 불변과 순수함수에 대해 말씀드렸는데요. 상황에 따라 불변을 유지하기 위한 비용이 오히려 더 큰 경우도 있습니다. 

무조건 불변과 순수함수를 택하기 보다는 가급적 불변과 순수함수를 지키는 형태로 구현해주시고 불변이나 순수함수로 구현했을 때 비용이 너무 크다면 가변을 사용하는 부분도 검토해보셔도 좋을 것 같아요.

wrallee replied:
잘 이해 된 것 같습니다. 답변 감사합니다😊

kyucumber commented on README.md:
```diff
@@ -10,3 +12,10 @@
 - [x] 사칙연산 기호가 아닌 경우 IllegalArgumentException throw
 - [x] 사칙 연산을 모두 포함하는 기능 구현
+
+### 자동차 경주
```
전체적으로 깔끔하게 잘 작성해주셨네요 👍

kyucumber reviewed (changes requested):
안녕하세요 몇가지 코멘트 남겨두었으니 확인 부탁드립니다.

wrallee committed:
refactor(RacingCar): 이동 계수 클래스 이름 변경(Score -> MoveFactor)

이동을 판단하기 위해 사용되는 변수이므로 Score보다는 MoveFactor로 변경

wrallee committed:
refactor(RacingCar): 애플리케이션, 서비스 이름 변경

RacingCarApplication.kt -> RacingGameApplication.kt
CarGame.kt -> RacingGame.kt

kyucumber reviewed (approved):
남긴 코멘트에 답변 달아두었으니 확인 부탁드립니다~

https://github.com/next-step/kotlin-racingcar/pull/817#discussion_r1023804856

다음 단계 진행을 위해 머지할게요
